TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -624,35 +734,63 @@ public void UpdateMissingCopies(Uri fileUri, List<string> RemainingMissingCopies
+                    ScheduleRefresh(docContext.Project);
+                }
             }
         }
 
         /// <summary>
-        /// Refresh all opened files' parser.
+        /// Refresh all opened files in all WorspaceProject instances
+        /// <remarks>This method is designed to run on background worker thread as it uses compilation results to recreate source code.
+        /// Callers must enqueue a new message to schedule a refresh of opened files instead of calling this directly.</remarks>
         /// </summary>
-        public void RefreshOpenedFiles()
+        private void RefreshOpenedFiles()
         {
             RefreshCustomSymbols();
-            this.CompilationProject.ClearImportedCompilationDocumentsCache();
+            this.WorkspaceProjectStore.RefreshOpenedFiles();
+        }
 
-            lock (_lockForOpenedDocuments)
-            {
-                var tempOpeneContexts = new Dictionary<Uri, DocumentContext >(_openedDocuments);
-                foreach (var contextEntry in tempOpeneContexts)
-                {
-                    Uri uri = contextEntry.Key;
-                    DocumentContext docContext = contextEntry.Value;
-                    var sourceText = new StringBuilder();
-                    foreach (var line in docContext.FileCompiler.TextDocument.Lines)
-                        sourceText.AppendLine(line.Text);
+        /// <summary>
+        /// Refresh One opened document
+        /// </summary>
+        /// <param name="docContext">The Document context to be refreshed</param>
+        /// <param name="clearCopyCache">true if the imported document cache must be cleared, false otherwise</param>
+        private void RefreshOpenedDocument(DocumentContext docContext, bool clearCopyCache)
+        {
+            if (clearCopyCache)
+                docContext.FileCompiler.CompilationProject.ClearImportedCompilationDocumentsCache();
+
+            var sourceText = new StringBuilder();
+            var compilationResults = docContext.FileCompiler?.CompilationResultsForProgram;
+            if (compilationResults == null)
+                return;
+            foreach (var cobolTextLine in compilationResults.CobolTextLines)
+                sourceText.AppendLine(cobolTextLine.Text);
 
-                    //Disconnect previous LanguageServer connection
-                    docContext.LanguageServerConnection(false);                    
+            //Disconnect previous LanguageServer connection
+            docContext.LanguageServerConnection(false);
 
-                    OpenTextDocument(docContext, sourceText.ToString(), LsrTestingOptions.NoLsrTesting);
+            BindFileCompilerSourceTextDocument(docContext, sourceText.ToString(), LsrTestingOptions.NoLsrTesting);
+        }
+
+        /// <summary>
+        /// Refresh all documents of a WorkspaceProject instance
+        /// </summary>
+        internal void ScheduleRefresh(WorkspaceProject project)
```

Comment:
I imagined a method `internal void ScheduleRefresh(List<DocumentContext> docs)` but you're right, as we have to clear the copy cache, we'll always have to refresh at least at WorkspaceProject level.

Now there is a problem with the current code, the `ScheduleRefresh` is intended to replace `RefreshOpenedFiles` but here you kept RefreshOpenedFiles which is passed to copy watchers, which in turn use an Action to wrap the call. The refresh is wrapped two times.